File: Graduation_Project_BE/api-customer/src/main/java/com/project/app/transaction/service/TransactionService.java
package com.project.app.transaction.service;

import com.project.app.transaction.dto.request.SePayWebhookRequest;
import com.project.app.transaction.dto.request.TopUpRequest;
import com.project.app.transaction.dto.response.TopUpResponse;
import com.project.app.transaction.entity.Transaction;
import com.project.app.user.entity.User;

import com.project.app.transaction.dto.request.TransferRequest;
import com.project.app.transaction.dto.response.TransferResponse;
import com.project.app.transaction.dto.request.WithdrawRequest;
import com.project.app.transaction.dto.request.ManualTransactionRequest;
import com.project.app.transaction.dto.response.WithdrawResponse;
import com.project.app.transaction.dto.response.ManualTransactionResponse;

public interface TransactionService {
    TopUpResponse initiateTopUp(User user, TopUpRequest request);
    void processSePayWebhook(String apikey, SePayWebhookRequest request);
    Transaction getTransactionByCode(String transactionCode, User user);
    Transaction updateTransaction(String transactionCode, User user, com.project.app.transaction.dto.request.UpdateTransactionRequest request);
    WithdrawResponse processWithdrawal(User user, WithdrawRequest request);
    TransferResponse processTransfer(User user, TransferRequest request);
    ManualTransactionResponse createManualTransaction(User user, ManualTransactionRequest request);
}

File: Graduation_Project_AppCustomer/shared/api/services/transactionService.ts
import { ENDPOINTS } from '../endpoints';
import { axiosClient } from '../axiosClient';

export interface TopUpRequest {
  walletId?: number;
  amount?: number;
}

export interface TopUpResponse {
  transactionCode: string;
  transferContent?: string;
  qrUrl?: string;
  expiresAt?: string;
  amount?: number;
  createdAt: string;
}


export interface TransferRequest {
  receiverAccountNumber: string;
  amount: number;
  pinCode: string;
  note?: string;
}

export interface TransferResponse {
  transactionCode: string;
  status: string;
  amount: number;
  receiverName: string;
  createdAt: string;
}

export interface WithdrawRequest {
  amount: number;
  bankAccountId: number;
  pinCode: string;
  note?: string;
}

export interface WithdrawResponse {
  transactionCode: string;
  status: 'PENDING' | 'SUCCESS' | 'FAILED' | 'CANCELLED';
  amount: number;
  createdAt: string;
}

export interface TransactionHistoryItem {
  transactionCode: string;
  type: string;
  status: string;
  amount: number;
  note?: string;
  categoryId?: number;
  categoryLabel?: string;
  categoryIcon?: string;
  categoryColor?: string;
  categoryDeleted?: boolean;
  createdAt: string;
}

export const transactionService = {
  initiateTopUp: async (data: TopUpRequest): Promise<TopUpResponse> => {
    const response = await axiosClient.post(ENDPOINTS.TRANSACTION.TOP_UP, data);
    return response.data;
  },

  
  internalTransfer: async (data: TransferRequest): Promise<TransferResponse> => {
    const response = await axiosClient.post(ENDPOINTS.TRANSACTION.TRANSFER, data);
    return response.data;
  },

  processWithdrawal: async (data: WithdrawRequest): Promise<WithdrawResponse> => {
    const response = await axiosClient.post(ENDPOINTS.TRANSACTION.WITHDRAW, data);
    return response.data;
  },

  getTransactionHistory: async (wallet: 'main' | string = 'main'): Promise<TransactionHistoryItem[]> => {
    const response = await axiosClient.get(ENDPOINTS.HISTORY.TRANSACTIONS, {
      params: { wallet },
    });
    return response.data || [];
  },

  updateTransaction: async (code: string, data: { categoryId?: number; note?: string }): Promise<any> => {
    const response = await axiosClient.put(`/api/v1/transactions/${code}`, data);
    return response.data;
  },
};
